rfcs/feature-set-metadata-file-to-web-platform-tests.md

```diff
@@ -0,0 +1,220 @@
+# RFC #163: Add web_feature metadata file to web-platform-tests
+
+Author: @jcscottiii
+
+# Introduction
+
+The web-platform-tests (WPT) project is a valuable resource for testing web
+platform features and contains valuable metadata such as feature spec links.
+Recently, the [WebDX Community Group](https://www.w3.org/community/webdx/)
+began creating the [web-features](https://github.com/web-platform-dx/web-features)
+repository. That repository serves as a basic shared catalog of feature
+definitions of the web platform. The repository itself doesn't intend to produce
+data but rather link to existing data that will inform audiences which features
+are part of [Baseline](https://web.dev/baseline/). However, there's a great
+opportunity to connect the WPT ecosystem to the web-features catalog. By doing
+this, it would enable users of wpt.fyi to filter by web-features grouping, which
+is similar to
+[the ability to filter by spec links](https://github.com/web-platform-tests/wpt.fyi/issues/1489).
+The RFC proposes the addition of a WEB_FEATURE.yml metadata file which would
+enable the linkage between WPT and web-features.
+
+# Proposed change
+
+The proposed change includes:
+
+1. Introduce a new metadata file, WEB_FEATURE.yml
+2. Add linting functionality
+3. Create a script to generate a manifest
+4. Adjust the wpt-pr-bot to handle reviews of the changes
+
+Below are the details for each step.
+
+## Step 1. Introduce a new metadata file, WEB_FEATURE.yml
+
+Originally, there was a RFC that proposed adding these details to the existing
+META.yml. During the August 01, 2023 Monthly WPT
+[Meeting](https://github.com/web-platform-tests/wpt-notes/blob/master/minutes/2023-08-01.md#rfc-157---add-feature-meta-tag-to-web-platform-tests-meta-data),
+it was discussed that it should be a separate file. This section describes the structure of the file.
+
+This file is expected to be in the same places developers would expect a META.yml.
+
+### File examples
+
+Typical example:
+
+```
+web_feature: subgrid
+```
+
+An example using all of the fields:
+
+```
+apply_mode: FORCE_RECURSIVE
```

Comment:
So generally these `apply_mode` / `override_mode` settings feel a bit over-engineered / action at a distance to me.

I'm also slightly concerned that this is different to the way that metadata ini files work and different to the way that the wpt-metadata repository works; it seems like each time we come across a similar problem we invent a slightly different solution.

Would it be enough to just have something like:

```
features:
   name: feature1
   files: *
```
For something that applies to all files (recursively) and then
```
features:
  name: feature2
  files:
    - test.html
    - subdir
 ```
 For features that only apply to specific files or directories, and then have a rule that we only inherit in cases that there's no more-specific rule in a subdirectory? And if we need multiple features per test (which I'm not sure if it's a requirement or not) we turn the `features` key into a list, and just say that if you want an inherited feature and a new one you need to repeat the inherited rule? Or inherit and add by default but have a magic feature name like `@reset` that blocks inherited features from applying (that's basically how this works in the expectation metadata).

Reply:
I like the idea of switching `features` to a list
I can see getting rid of `override_mode`. But I think the `apply_mode` field makes it clearer.

Default case:

```yaml
features:
   - name: feature1
     files: **
```

List out files case:

```yaml
features:
- name: test-2
  files:
  - test-2-01.html
  - test-2-02.html
- name: test-1
  files:
  - test-1-01.html
```

Ignore parent features (or `@reset`)

```yaml
apply_mode: IGNORE_PARENT
features:
- name: reset-test-1
  files: "**"
```

Ignore parent features (or `@reset`) while listing out explicit files

```yaml
apply_mode: IGNORE_PARENT
features:
- name: reset-test-2
  files:
  - reset-test-2-01.html
- name: reset-test-1
  files:
  - reset-test-2-01.html
```

The jsonschema would look like: 

```json
{
    "$schema": "http://json-schema.org/draft-06/schema#",
    "$ref": "#/definitions/Main",
    "definitions": {
        "Main": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "features": {
                    "type": "array",
                    "description": "List of features",
                    "items": {
                        "$ref": "#/definitions/FeatureEntry"
                    }
                },
                "apply_mode": {
                    "type": "string",
                    "anyOf": [
                    	{
                            "const": "IGNORE_PARENT",
                            "description": "Ignores features from previous parent directories."
                        }
                    ]
                }
            },
            "required": [
                "features"
            ],
            "title": "Main"
        },
        "FeatureEntry": {
            "type": "object",
            "additionalProperties": false,
            "properties": {
                "name": {
                    "type": "string",
                    "description": "The web feature key"
                },
                "files": {
                    "oneOf": [
                        {
                            "const": "**",
                            "description": "All files recursively",
                            "title": "Files"
                        },
                        {
                            "type": "array",
                            "items": {
                                "type": "string",
                                "description": "A specific file within the current directory"
                            }
                        }

                    ]
                }
            },
            "required": [
                "name",
                "files"
            ],
            "title": "FeatureEntry"
        }
    }
}
```

Thoughts @jgraham ? If this looks fine, I will readjust the RFC.

Also, I did some experimentation on my fork ([script](https://github.com/jcscottiii/rfcs/blob/experiment/test.py)) and [testdata directory](https://github.com/jcscottiii/rfcs/tree/experiment/testdata).

<details>
  <summary>Result</summary>
  
  ```
FeatureResult(_feature_tests_map_: OrderedDict([('root-feature', {'testdata/feature-policy/resources/feature-policy-picture-in-picture.html', 'testdata/css/grid/subgrid/subgrid1.html', 'testdata/html/html-test.html', 'testdata/picture-in-picture/pip1.html', 'testdata/css/grid/grid1.html', 'testdata/css/grid/grid2.html'}), ('picture-in-picture', {'testdata/picture-in-picture/pip1.html'}), ('grid', {'testdata/css/grid/subgrid/subgrid1.html', 'testdata/css/grid/grid1.html', 'testdata/css/grid/grid2.html'}), ('subgrid', {'testdata/css/grid/subgrid/subgrid1.html'}), ('reset-test-1', {'testdata/html/reset-test-1/reset-test-1-01.html', 'reset-test-2-01.html', 'testdata/html/reset-test-1/reset-test-1-02.html'}), ('reset-test-2', {'reset-test-2-01.html'})]))
  ```
</details>

Reply:
A `features` list makes sense to me for cases where one directory has tests for multiple features, which is the case for https://wpt.fyi/results/css/selectors for example. In some of those cases it would be nice to just group them into directories, but I doubt we can *always* do it and thus avoid the problem entirely.

FWIW, I don't think it's an important use case to have one test be associated with multiple features, but if it's incidentally supported that's fine.

Regarding inheritance, I think we ought to support marking up all of https://wpt.fyi/results/css/css-grid *except* the masonry/ and subgrid/ subdirectories as being part of the grid feature.

Reply:
As discussed in today's wpt infra [meeting](https://github.com/web-platform-tests/wpt-notes/blob/master/minutes/2023-10-03.md), I added the new jsonschema from above into the RFC.